pkg/library/home/persistentHome.go

```diff
@@ -95,3 +99,77 @@ func storageStrategySupportsPersistentHome(workspace *common.DevWorkspaceWithCon
 	storageClass := workspace.Spec.Template.Attributes.GetString(constants.DevWorkspaceStorageTypeAttribute, nil)
 	return storageClass != constants.EphemeralStorageClassType
 }
+
+func addInitContainer(dwTemplateSpec *v1alpha2.DevWorkspaceTemplateSpec, initContainerFromConfig *v1alpha2.Container) {
+
+	initContainerComponent := v1alpha2.Component{
+		Name: constants.HomeInitComponentName,
+	}
+
+	initContainer := initContainerFromConfig
+
+	if initContainer == nil {
+		initContainer = inferInitContainer(dwTemplateSpec)
+	}
+
+	if initContainer == nil {
+		return
+	}
+
+	initContainerComponent.ComponentUnion = v1alpha2.ComponentUnion{
+		Container: &v1alpha2.ContainerComponent{
+			Container: *initContainer,
+		},
+	}
+
+	dwTemplateSpec.Components = append(dwTemplateSpec.Components, initContainerComponent)
+	dwTemplateSpec.Commands = append(dwTemplateSpec.Commands, v1alpha2.Command{
+		Id: constants.HomeInitEventId,
+		CommandUnion: v1alpha2.CommandUnion{
+			Apply: &v1alpha2.ApplyCommand{
+				Component: constants.HomeInitComponentName,
+			},
+		},
+	})
+	dwTemplateSpec.Events.PreStart = append(dwTemplateSpec.Events.PreStart, constants.HomeInitEventId)
+}
+
+func inferInitContainer(dwTemplateSpec *v1alpha2.DevWorkspaceTemplateSpec) *v1alpha2.Container {
+	var nonImportedComponent v1alpha2.Component
+	for _, component := range dwTemplateSpec.Components {
+		if component.Container == nil {
+			continue
+		}
+
+		pluginSource := component.Attributes.GetString(constants.PluginSourceAttribute, nil)
+		if pluginSource == "" || pluginSource == "parent" {
+			// First, non-imported container component is selected
+			nonImportedComponent = component
+			break
+		}
+	}
+
+	createOutFile := "touch /home/user/.init-persistent-home"
+	runSetupScript := `([ -n "$HOME_SETUP_SCRIPT" ] && [ -f "$HOME_SETUP_SCRIPT" ] && echo "Running $HOME_SETUP_SCRIPT" && (source "$HOME_SETUP_SCRIPT" || true))`
```

Comment:
This `$HOME_SETUP_SCRIPT` is supposed to be an environment variable that points to a user-configured script for configuring home-persistence, right? For cases where users may be using an image other than the UDI and the default `runStow` script for home-persistence won't work. 

And to make sure I understand the logic here: we first check if the user defined the `$HOME_SETUP_SCRIPT` environment variable, then ensure it points to an actual file, then we run it using `source` and ensure it does not cause workspace startup to fail with `|| true`. Correct?

Reply:
> This $HOME_SETUP_SCRIPT is supposed to be an environment variable that points to a user-configured script for configuring home-persistence, right? 

Right, it allows users to provide their own script in the image.

> and ensure it does not cause workspace startup to fail with || true. Correct?

Yes, the `|| true` is to ensure that the initcontainer command's exit code is 0 allowing the workspace to still start if there was a non-zero exit code in the initialization script.